feat: Java bindings moved from Velox4J

zhztheplayer opened the pull request:
## Goal

Goal of the PR is to add complete Java bindings to Velox. The code is migrated from https://github.com/velox4j/velox4j. After this PR is landed, the Velox4J repository will be archived.

The architecture of the Java bindings code is designed to be general, to support different upper layer JVM applications.

## Background

The previous related discussions:

1. The formal discussion about the contribution of Velox4J as part of Velox: https://github.com/facebookincubator/velox/discussions/12808
2. (Apache Gluten) The initial PR to add a Flink accelerator which is backed by Velox4J and Velox (author @shuai-xu): https://github.com/apache/incubator-gluten/pull/8839

### About Velox4J

For more information about the dedicated project Velox4J, see its latest readme: https://github.com/velox4j/velox4j/blob/main/README.md.

## This PR

### Introduction

A new readme for the Java bindings is also being added: https://github.com/zhztheplayer/velox/blob/wip-java-bindings-2/java/README.md. The readme file was modified from Velox4J's readme.

### Directory Structure

The new folder source `java` will be added to Velox source base for the Java bindings code. The following figure represents the code structure:

```
velox
|- .github
|    \- workflows
|        |- ...
|        |- java-format.yml (ADDED)
|        |- java-build-java.yml (ADDED)
|        \- java-build-cpp.yml (ADDED)
|- velox
|- python
|- ...
\- java (ADDED)
    |- velox4j (ADDED)
       \- ... (ADDED)
    \- README.md (ADDED)
```

### Supported Velox Features

The following lists the already-supported Velox features as of this PR in the Java bindings.

#### Data Types
All data types of Velox are supported by the Java bindings in this PR.

- ArrayType
- BigIntType
- BooleanType
- DateType
- DecimalType
- DoubleType
- FunctionType
- HugeIntType
- IntegerType
- IntervalDayTimeType
- IntervalYearMonthType
- MapType
- OpaqueType
- RealType
- RowType
- SmallIntType
- TimestampType
- TinyIntType
- VarbinaryType
- VarCharType

#### Operators
The following Velox plan nodes are supported by the Java bindings in this PR.

- TableScanNode
- TableWriteNode
- ProjectNode
- OrderByNode
- LimitNode
- HashJoinNode
- FilterNode
- AggregationNode

#### Expressions
The following Velox expressions are supported by the Java bindings in this PR.

- CallTypedExpr (with Spark functions registered by default)
- CastTypedExpr
- ConcatTypedExpr
- ConstantTypedExpr
- DereferenceTypedExpr
- FieldAccessTypedExpr
- InputTypedExpr
- LambdaTypedExpr

#### File Formats
All file formats supported in Velox are supported by the Java bindings in this PR.

- parquet
- orc
- nimble
- json
- text
- dwrf
- rc
- sst

#### Variants (Constants)
All variants supported in Velox are supported by the Java bindings in this PR.

#### Plan Execution
Currently only serial (single-threaded) execution is supported by the Java bindings in this PR.

#### Expression Evaluation
Simple Velox expression evaluation, without creating a Velox plan node, is supported with class `Evaluator`. 

### Preset
Preset that can be configured with Velox4J configuration `velox4j.init.preset` during initialization.

Preset affects connector registration and function registration. So far only the preset for Spark can be specified. The API is extensible for adding new presets in future.

### Build

Build requires x86 + Linux platform as of now.

#### Build C++ code only

```shell
cd java/velox4j
src/main/cpp/build.sh
```

#### Build bundle with both Java and C++ code

```shell
cd java/velox4j
mvn clean install -Dgpg.skip
```

### Testing
There are 100+ UT cases being added, covering core architecture, plan serialization / deserialization, query execution, expression evaluation, arrow support, and more. 

#### Run Java UTs:

```shell
cd java/velox4j
mvn test -Dgpg.skip
```

#### Run C++ UTs:

```shell
cd java/velox4j/src/main/cpp
./test.sh
```

### CI

Currently in this PR we are adding 3 CI workflows to Velox's GHA CI for the Java bindings:

- java-format.yml (for every PR, runs format check for Java bindings' code)
- java-ut-cpp.yml (daily, runs C++ UTs for the C++ code in Java bindings)
- java-ut-java.yml (for every PR, runs Java UTs for Java bindings)

### Releases

Ideally, the Java bindings code can be released as Maven package `com.meta.velox4:velox4j`. But this will involve in Meta's internal process which can be figured out in future.

As of now, user is required to build the Java bindings (Velox4J) by themselves. Following the [readme](https://github.com/zhztheplayer/velox/blob/wip-java-bindings-2/java/README.md#build-from-source) to create a local Maven build.

zhztheplayer committed:
contrib velox4j

zhztheplayer commented on .github/workflows/java-format.yml:
```diff
@@ -0,0 +1,56 @@
+      - uses: actions/checkout@v4
+      - name: Run format checker
+        run: |
+          cd java/velox4j
+          bash scripts/gha/format/format.sh -check
```
So far `/java/velox4/scripts/gha/format/format.sh` is used as linter for Java bindings code specifically. We could integrate it into the main `make` formatter in the future.

zhztheplayer commented on .github/workflows/java-ut-java.yml:
```diff
@@ -0,0 +1,113 @@
+      - "third_party/**"
+      - "scripts/setup-ubuntu.sh"
+      - "scripts/setup-helper-functions.sh"
+      - "java/velox4j/**"
+      - ".github/workflows/java-ut-java.yml"
```
We add the CI path triggers for Java code in addition to the triggers for C++.

zhztheplayer commented on java/velox4j/src/main/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+  set(VELOX_ENABLE_CCACHE OFF)
+endif()
+
+# Import Velox.
+set(VELOX_ROOT_DIR "${PROJECT_SOURCE_DIR}/../../../../../")
+FetchContent_Declare(
+  velox
+  DOWNLOAD_COMMAND
+    ln -sf ${VELOX_ROOT_DIR}/velox ${FETCHCONTENT_BASE_DIR}/velox-src/velox &&
+    ln -sf ${VELOX_ROOT_DIR}/scripts ${FETCHCONTENT_BASE_DIR}/velox-src/scripts
+    && ln -sf ${VELOX_ROOT_DIR}/CMake ${FETCHCONTENT_BASE_DIR}/velox-src/CMake
+    && ln -sf ${VELOX_ROOT_DIR}/CMakeLists.txt
+    ${FETCHCONTENT_BASE_DIR}/velox-src/CMakeLists.txt)
+FetchContent_MakeAvailable(velox)
```
The core velox lib core is imported through `FetchContent`.

To avoid the recursive dependency `java -> velox -> java` mentioned in https://github.com/facebookincubator/velox/discussions/13044, we selectively create several symbolic links to these needed core velox files in java bindings' build dir:

- velox/velox
- velox/scripts
- velox/CMake
- velox/CMakeLists.txt

Symbolic links of these 4 items are enough to build the velox core lib as the dependency of java bindings' cpp code.

zhztheplayer replied:
> add_subdirectory(${VELOX_ROOT_DIR} ${VELOX_BINARY_DIR})

I once tried this but left out `${VELOX_BINARY_DIR}` so got some errors. Now it seems working fine with the binary dir set. Thanks for the suggestion. :+1:

zhztheplayer commented on java/velox4j/src/main/cpp/build.sh:
```diff
@@ -0,0 +1,74 @@
+  fi
+done
+
+# Force '$ORIGIN' runpaths for all so libraries to make the build portable.
+# 1. Remove any already set RUNPATH sections.
+for file in "$INSTALL_DESTINATION"/*
+do
+  echo "Removing RUNPATH on file: $file ..."
+  patchelf --remove-rpath "$file"
+done
+
+# 2. Add new RUNPATH sections with '$ORIGIN'.
+for file in "$INSTALL_DESTINATION"/*
+do
+  echo "Adding RUNPATH on file: $file ..."
+  patchelf --set-rpath '$ORIGIN' "$file"
+done
+
+# 3. Print new ELF headers.
+for file in "$INSTALL_DESTINATION"/*
+do
+  echo "Checking ELF header on file: $file ..."
+  readelf -d "$file"
+done
+
+# Do final checks.
+# 1. Check ldd result.
+echo "Checking ldd result of libvelox4j.so: "
+ldd "$INSTALL_DESTINATION/$VELOX4J_LIB_NAME"
+
+# 2. Check ld result.
+echo "Checking ld result of libvelox4j.so: "
+ld "$INSTALL_DESTINATION/$VELOX4J_LIB_NAME"
+
+# Finished.
+echo "Successfully built velox4j-cpp."
```
The `RUNPATH` modification is for all libraries in the installation folder, to make sure they find the needed dependencies from the current folder first. This is important for creating a portable C++ build to be carried into the Java fat jar.

zhztheplayer replied:
It looks here the problem could be a little bit more complicated because all the runtime dependencies of target `velox` will also need to be modified the rpaths. I tried with CMake property `INSTALL_RPATH` from the beginning but it didn''t work really well with the installation method we are using on Velox4J. 

Do you have any suggestions? And anyways it's definitely the best idea to integrate these actions into the CMake script. I could keep tracking on this topic.

zhztheplayer reviewed (commented):
Hi @assignUser, would you like to help me review the CI and build system of the Java bindings? Thanks!

zhztheplayer commented:
Also kindly ping @shuai-xu @zhanglistar @lgbo-ustc @KevinyhZou @PHILO-HE @JkSelf @rui-mo @jinchengchenghh @marin-ma from the Apache Gluten community for more code review.

Thank you for all your help in advance.

jinchengchenghh reviewed (commented):
Awesome work! Some initial comments. Do we always use the latest cpp velox version to combine with velox4j? If some one changes the velox cpp code, what's the efforts to change the java code?

zhztheplayer commented on java/velox4j/scripts/gha/common/setup-centos7.sh:
```diff
@@ -0,0 +1,122 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
```
I think Velox should support Ubuntu 24? https://github.com/facebookincubator/velox?tab=readme-ov-file#setting-up-on-ubuntu-2004-or-later

And you are right, CentOS 7 was deprecated by Velox already (it's a good question). We consider CentOS 7 here because it has a lower glibc version so we can ship the Jar to a wilder set of linux distributions. I think in the future we may provide official Jar releases that were built on low-glibc Linux either.

jinchengchenghh replied:
Do we need java/velox4j/scripts/gha/common/setup-ubuntu24.sh if we can run https://github.com/facebookincubator/velox?tab=readme-ov-file#setting-up-on-ubuntu-2004-or-later

zhztheplayer replied:
> Do we need java/velox4j/scripts/gha/common/setup-ubuntu24.sh if we can run https://github.com/facebookincubator/velox?tab=readme-ov-file#setting-up-on-ubuntu-2004-or-later

Some additional packages are needed by Velox4J but are not needed by Velox itself, for instance, `patchelf` and the openjdk. But you are right we may have possibility to use `velox/scripts/setup-ubuntu.sh` to simplify this script. May be something we can follow up with an individual PR?

zhztheplayer commented on java/velox4j/pom.xml:
```diff
@@ -0,0 +1,278 @@
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>com.diffplug.spotless</groupId>
```
1. spotless-maven-plugin
2. maven-checkstyle-plugin

They are actually functionally overlapped because they both do Java style checks. As known the former do style-fix as well. So we just keep one spotless plugin for simplicity.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/init/Init.cc:
```diff
@@ -0,0 +1,102 @@
+  f();
+}
+
+void initForSpark() {
```
> How can we handle the spark and presto

We provide the configuration `velox4j.init.preset` which is defined [here](https://github.com/zhztheplayer/velox/blob/98c9a22fe6b0e31e605ed113b51f5f14e159bc42/java/velox4j/src/main/cpp/main/velox4j/init/Config.cc#L5-L7). So far only Spark is supported but in the future we could have more.

> maybe remove spark specified code to a separate directory

I agreed. This can be a refactor in the future because we don't support Presto or Flink yet. This will involve some directory structure redesign but I don't have them clear in mind yet.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniCommon.cc:
```diff
@@ -0,0 +1,140 @@
+
+namespace velox4j {
+
+const jint kJniVersion = JNI_VERSION_1_8;
```
Will remove this var because it's actually unused. Thanks.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniWrapper.cc:
```diff
@@ -0,0 +1,421 @@
+using namespace facebook::velox;
+
+namespace {
+const char* kClassName = "com/meta/velox4j/jni/JniWrapper";
```
Explained by https://github.com/facebookincubator/velox/pull/13186#discussion_r2082029965, thanks.

zhztheplayer commented on java/velox4j/src/main/java/com/meta/velox4j/variant/VarCharValue.java:
```diff
@@ -0,0 +1,54 @@
+import com.fasterxml.jackson.annotation.JsonGetter;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class VarCharValue extends Variant {
```
It's possible but not necessarily worth it because we do have manual-writing code in the implementations to align with C++ variant definitions. Auto-generation of the code will require introducing a Java template engine tool, and the manual-writing code will still be there. So it will become over-engineering based on the demand so far.

zhztheplayer commented on java/velox4j/src/main/java/com/meta/velox4j/plan/TableScanNode.java:
```diff
@@ -0,0 +1,65 @@
+import com.meta.velox4j.connector.ConnectorTableHandle;
+import com.meta.velox4j.type.Type;
+
+public class TableScanNode extends PlanNode {
```
Similar to the above discussion https://github.com/facebookincubator/velox/pull/13186#discussion_r2082083787.

Moreover, some serialization protocols like protobuf, flatbuffers could provide code-gen tools because they are schema-aware. Schema can be used for generating code. But JSON is schema-less so we don't have a pre-defined protocol that can be used for code-gen.

zhztheplayer commented:
I've moved all non-source files (CI scripts, CMake files, build scripts, Maven POM files, etc.) to the [first commit](https://github.com/facebookincubator/velox/pull/13186/commits/ead26b78adf3bb2fdf4dbc6d520e7f12174704af) of this PR to ease review. Thanks! cc @assignUser

JkSelf commented on java/README.md:
```diff
@@ -0,0 +1,201 @@
+
+### Compatible With Arrow Java
+
+Velox4J is compatible with [Apache Arrow's Java implementation](https://arrow.apache.org/java/). Built-in utilities converting between
```
As discussed [here](https://github.com/facebookincubator/velox/discussions/12808#discussioncomment-12650164), introducing the Arrow data format can be costly. Could we consider adopting Gluten's implementation, which directly converts the Spark format to the Velox format, thereby bypassing the Arrow format? This approach might require applications to implement the format conversion themselves. However, it could lead to performance improvements. Alternatively, we could offer an API that allows users to implement the data conversion without directly using the Arrow format.

zhztheplayer replied:
> Could we consider adopting Gluten's implementation, which directly converts the Spark format to the Velox format, thereby bypassing the Arrow format?

This is feasible. These native Spark2Velox / Flink2Velox implementations can be added directly to Java bindings code based but with some extensible API designs. For Spark, Velox already provides the transition APIs so all we need is to expose them from C++ to Java.

JkSelf commented on java/velox4j/src/main/cpp/main/velox4j/init/Config.h:
```diff
@@ -0,0 +1,24 @@
+
+#include <velox/common/config/Config.h>
+
+namespace velox4j {
```
To maintain consistency with other modules such as `facebook::velox::py`, could we change the `velox4j `namespace to `facebook::velox::velox4j`?

zhztheplayer replied:
I'll prefer `facebook::velox4j` since it's not sharing the same project root with Velox C++?

JkSelf replied:
@zhztheplayer It appears that most of the Python modules are located in the `velox/python` directory https://github.com/facebookincubator/velox/tree/main/velox/python, with only a few test codes in the `/python `folder https://github.com/facebookincubator/velox/tree/main/python. If we aim to maintain Java at the same level as Python within Velox, should we also organize the functional code similarly by placing it in `velox/java` rather than directly in the `/java` directory?

zhztheplayer replied:
> should we also organize the functional code similarly by placing it in velox/java rather than directly in the /java directory?

Thank you but it's not a good idea from my perspective because it breaks the isolation between Java and C++. My intention was to encapsulate the Java bindings completely within a individual folder, development and maintenance will be made much easier because of doing that.

JkSelf replied:
@zhztheplayer It appears that the Java directory in this PR contains both Java and C++ code, rather than just Java code

zhztheplayer replied:
`java` stands for the Java bindings, it's not necessarily for placing Java code only.

JkSelf commented on java/README.md:
```diff
@@ -0,0 +1,201 @@
+The minimum toolchain versions for building Velox4J:
+
+- GCC 11
+- JDK 11
```
Are there plans to support JDK 8, considering that some customers might still be using it?

zhztheplayer replied:
It's a valid point to me. And I think it's not very difficult to add the support given we don't use much of Java 11 language features yet. Could be a topic for following up.

JkSelf commented on java/velox4j/src/main/java/com/meta/velox4j/connector/HiveSortingColumn.java:
```diff
@@ -0,0 +1,46 @@
+import com.meta.velox4j.serializable.ISerializable;
+import com.meta.velox4j.sort.SortOrder;
+
+public class HiveSortingColumn extends ISerializable {
```
Can we add new `hive `folder under `connector ` to put the Hive related code?

zhztheplayer replied:
let me check

JkSelf commented:
@zhztheplayer Thank you for your excellent work. The changes are quite extensive, making it challenging to review. Is there a possibility to split it into several smaller PRs to expedite the review process? Thanks.

zhztheplayer commented:
> @zhztheplayer Thank you for your excellent work. The changes are quite extensive, making it challenging to review. Is there a possibility to split it into several smaller PRs to expedite the review process? Thanks.

I was considering something similar but it's uncertain how small the PRs will be because the feature is a whole, so I wonder it might not help a lot but will cost me a lot of energy.

Do you think it's helpful to view Velox4J's commit history to find how the project was iterated back then? 

https://github.com/velox4j/velox4j/commits/main/

Though the first commit was already large enough with 10000+ code. The majority of the code was for building the base architecture so it's not that easy to break it into small snippets.

JkSelf commented:
> > @zhztheplayer Thank you for your excellent work. The changes are quite extensive, making it challenging to review. Is there a possibility to split it into several smaller PRs to expedite the review process? Thanks.
> 
> I was considering something similar but it's uncertain how small the PRs will be because the feature is a whole, so I wonder it might not help a lot but will cost me a lot of energy.
> 
> Do you think it's helpful to view Velox4J's commit history to find how the project was iterated back then?
> 
> https://github.com/velox4j/velox4j/commits/main/
> 
> Though the first commit was already large enough with 10000+ code. The majority of the code was for building the base architecture so it's not that easy to break it into small snippets.

@zhztheplayer Could we split them into several PRs based on the supported features you listed? These categories could include:
```
Data Types
Operators
Expressions
File Formats
Variants (Constants)
Build
Test
CI
```
This approach might streamline the review process.

zhztheplayer committed:
rebase

zhztheplayer committed:
rebase

lgbo-ustc commented on java/velox4j/src/main/cpp/main/velox4j/eval/Evaluator.h:
```diff
@@ -0,0 +1,40 @@
+      const facebook::velox::RowVector& input);
+
+ private:
+  MemoryManager* const memoryManager_;
```
Should `memoryManager_` and `exprJson_` be a member variable if they're only used in the constructor?

lgbo-ustc commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/ObjectStore.h:
```diff
@@ -0,0 +1,122 @@
+  ObjectHandle save(std::shared_ptr<T> obj) {
+    const std::lock_guard<std::mutex> lock(mtx_);
+    const std::string_view description = typeid(T).name();
+    ResourceHandle handle = store_.insert(std::move(obj));
```
`store_` is thread-safe. We may move the lock acquisition from line 85 to behind line 87.

zhztheplayer replied:
Good question! It was actually intentional to keep the joint operations on `store_` and `aliveObjects_` atomic. 

See the `release` function:

https://github.com/facebookincubator/velox/pull/13186/files/0d9a73c07803b7aa40aa12754f44676ff3265363#diff-acbc62f2037e212eb63461a33d511eba6d07a304453cd66aca7db054c9cf6579R30-R34

So we have to have `store_` and `aliveObjects_` synchronized at a time, or else inconsistency might be caused when writing and reading happen concurrently.

zhztheplayer committed:
fixup

zhztheplayer committed:
fixup

lgbo-ustc commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cc:
```diff
@@ -0,0 +1,387 @@
+velox::memory::MemoryPool* MemoryManager::getVeloxPool(
+    const std::string& name,
+    const velox::memory::MemoryPool::Kind& kind) {
+  if (veloxPoolRefs_.count(name) > 0) {
+    const auto& pool = veloxPoolRefs_[name];
+    VELOX_CHECK_EQ(
+        pool->kind(),
+        kind,
+        "Pool [" + name + "] was already created but with different kind");
+    return pool.get();
+  }
```
I prefer to write it in the following way.
```c++
if (auto it = veloxPoolRefs_.find(name); it != veloxPoolRefs_.end()) {
     const auto & pool = iter->second;
     return pool.get();
}
```

zhztheplayer committed:
fixup

zhztheplayer committed:
fixup

zhztheplayer committed:
fixup

zhztheplayer commented:
Two high-level questions that may require help:

1. Java package name: `com.meta.velox4j` (this PR) vs `com.facebook.velox4j`, or any other ideas?
2. C++ code namespace name `velox4j::` (this PR) vs `facebook::velox::java::` vs `facebook::velox4j::`, or any other ideas?

cc @pedroerp @assignUser

zhztheplayer committed:
Package / namespace changes

Change Java package to `com.facebook.velox4j`
Change C++ namespace to `facebook::velox4j`

zhztheplayer committed:
Pom fixes

zhztheplayer committed:
fixup

zhztheplayer committed:
fixup

zhztheplayer committed:
empty

zhztheplayer committed:
empty

assignUser reviewed (commented):
Thanks for this big effort!

I did a first pass and left a few comments, sorry that it took me so long to get around to it :)

>C++ code namespace name velox4j:: (this PR) vs facebook::velox::java:: vs facebook::velox4j::, or any other ideas?

python uses `facebook::velox::py` so `facebook::velox::java` seems more aligned but I don't really have a strong opinion.

zhztheplayer committed:
Sync https://github.com/velox4j/velox4j/commit/860d09f012cbe6d90a1cb35e77fbeee88142d99a

zhztheplayer commented on .github/workflows/java-format.yml:
Thank you for the reminder. I could help to check how to make the hook work with Velox4J later on, then we can remove the linter routine in Velox4J.

assignUser replied:
As I said offline, lets do this as a follow up to not further blow up the scope of this pr.

zhztheplayer committed:
address comments

zhztheplayer commented on .github/workflows/java-ut-cpp.yml:
```diff
@@ -0,0 +1,53 @@
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: "0 8 * * *"
```
Yes, I observed the same in my own repositories. Do you have any suggestions for it? May be we can try sticking the cron to a specific minute like "43 8 * * *" or something?

assignUser replied:
Yes, that's what I do. Though you should stay a way from meme numbers like 69, 42, 23 :D

zhztheplayer commented on .github/workflows/java-ut-cpp.yml:
```diff
@@ -0,0 +1,53 @@
+# limitations under the License.
+name: Java Bindings - C++ UT
+
+on:
```
Ah, the `schedule` trigger haven't yet provided a `path` sub-trigger: https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#schedule

zhztheplayer commented on .github/workflows/java-ut-cpp.yml:
```diff
@@ -0,0 +1,53 @@
+            java-ut-cpp-ut-centos7-ccache-
+      - name: Run setup script / Build and run UTs
+        run: |
+          docker run -v ${{ github.workspace }}:/velox -w / centos:7 bash -c "
```
The script is simple as of now because the preparations are all bundled in `setup-centos7.sh` / `setup-ubuntu24.sh`. 

Would you suggest having a docker file and put a `RUN setup-centos7.sh` line in it? If we don't push the image to dockerhub it won't help much here and will slightly add some complexity to the GHA script from my view. Or you meant another approach?

zhztheplayer commented on java/velox4j/src/main/cpp/.clang-format:
I got what your meant. Though will it sound good to have an individual patch after the PR is merged, to enable `pre-commit` for Velox4J and to remove the original linter toolchains for it?

zhztheplayer commented on java/velox4j/src/main/cpp/build.sh:
```diff
@@ -0,0 +1,74 @@
+do
+  if [ -L "$file" ]
+  then
+    echo "CMake installation just created a symbolic link $file in the installation directory, this is not expected, aborting..."
```
This used to happen on some Linux dists when Velox4J was relying on CMake default installation logic at the beginning. And it should no longer exist but could be kept as a guard to ensure the sanity of the following statements.

zhztheplayer commented on java/velox4j/src/main/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+FetchContent_MakeAvailable(velox)
+
+# Import JniHelpers.
+set(JNI_HELPERS_REPO zhztheplayer/JniHelpersCpp)
```
I'll try to find a solution right in this PR. Maybe the original project can be imported as a whole.

zhztheplayer committed:
iwyu

zhztheplayer committed:
fixup

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/conf/Config.h:
```diff
@@ -0,0 +1,65 @@
+class ConfigArray : public facebook::velox::ISerializable {
+ public:
+  explicit ConfigArray(
+      std::vector<std::pair<std::string, std::string>>&& values)
```
fixed for all files

zhztheplayer committed:
Move to upstream spotify JniHelpers

zhztheplayer committed:
Change the C++ Build workflow schedule time to `38 8 * * *`

zhztheplayer commented on java/velox4j/scripts/gha/common/setup-ubuntu24.sh:
```diff
@@ -0,0 +1,44 @@
+apt-get install -y autoconf automake g++ libnuma-dev libtool numactl unzip libdaxctl-dev
+apt-get install -y openjdk-11-jdk
+apt-get install -y maven cmake
+apt-get install -y chrpath patchelf
```
fixed

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,69 @@
+/*
```
I wonder why we need the `java` top-level directory because it looks a bit strange to have cpp file under it. Can we just start with `velox4j`?

rui-mo replied:
There's a subdirectory `test` under python, while under `java` there's only the `velox4j` folder which might make the layer appear unnecessary. I wonder if we can follow the similar structure as `python`.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,69 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "Arrow.h"
```
Please use full include path for the headers in Velox.

rui-mo replied:
Perhaps use "velox4j/arrow/Arrow.h"?

zhztheplayer replied:
Ahh, thanks. Will change to follow the style.

rui-mo reviewed (commented):
Thanks for this tremendous work! Added some comments.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,69 @@
+ */
+#include "Arrow.h"
+
+#include <velox/vector/ComplexVector.h>
```
I got what you mean, though Velox4J considers Velox as an external dependency both by the separated-directory design and in the build system:

https://github.com/facebookincubator/velox/pull/13186/files#diff-788c4b0f67ce801e6d724cf17bd4c836851d301fa11d573c27f936c96936e23dR42-R47

So let's just keep `<velox/...>` for clearance? So people will easily distinguish them with the velox4j headers.

zhztheplayer replied:
Velox's practically the external dependency of Velox4J. The headers and the libraries are imported into Velox4J's build routine just like a regular 3rd dependency. In that case, "<" is the best choice.

> since this code will become part of Velox, we need to adhere to Velox's code style.

It's not only about the coding style. In most compilers, `#include "..."` is basically a superset of `#include <...>`. For making the code least functional, it's a good practice to do `#include <...>` if we can.

rui-mo replied:
From my view, the main issue here is not with the compilation process but rather with the way velox4j is intended to be integrated with Velox.  If we insist that Velox is the external dependence of Velox4J from the high-level design, It is unclear to me why this code should belong to Velox. If not, I think it's odd to claim that something in Velox isn't truly Velox.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,69 @@
+    ArrowSchema* cSchema,
+    ArrowArray* cArray) {
+  auto options = makeOptions();
+  return importFromArrowAsOwner(*cSchema, *cArray, pool);
```
> fromArrowToBaseVector

👍 Thank for catching. Will remove.

> this function might could be replaced with importFromArrowAsOwner.

Let's not do such simplification atm. Having an uniform entrypoint of code could help us on changing the behaviour in the future in one go.

rui-mo replied:
As for now, I think we don't need this function which just calls another function. 
> changing the behaviour in the future in one go

I'm not sure how the behaviour would be changed in the future, would you please clarify a bit?

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/connector/ExternalStream.h:
```diff
@@ -0,0 +1,191 @@
+      folly::Executor* cpuExecutor) override;
+};
+
+} // namespace facebook::velox4j
```
Thank you for pointing this out.

Would you mind if we have an individual PR for enhancing the code documents after this is merged. We'd prioritize the public APIs exposed from Java over these internal APIs. (But definitely not all the classes and members... I assume you mainly meant the public APIs :) )

So far the APIs are straightforward enough to use for users - our Flink developers are using / developing Velox4J in deep without directly engaging with me, so I assume the current documentation is not too little to block this PR's progress. Let alone we have the detailed PR description and a comprehensive readme. But we are exactly on the same page the code documentation should be enriched for public APIs and I will work on that soon.

rui-mo replied:
Could you please review the code style requirement first? I don't think we can land this PR without following it. Still, I would suggest separate this PR into several patches to help the review and merge process. Otherwise, a tiny comment (like this one) would yield to large efforts from the developer.

https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#comments

zhztheplayer replied:
> Still, I would suggest separating this PR into several patches to help the review and merge process. 

I high recommend considering something like this when we have no other ways. The code was moved from Velox4J to here but not written from scratch., separating everything again is a huge effort. It's in a whole folder and stand for one single topic (the Java bindings). 

I could add docs just in this PR though.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/init/Config.cpp:
```diff
@@ -0,0 +1,23 @@
+using namespace facebook::velox;
+config::ConfigBase::Entry<Preset> VELOX4J_INIT_PRESET(
+    "velox4j.init.preset",
+    Preset::SPARK);
```
Umm... `Preset::SPARK` is the default value for this option but not hard-coded.

Though I feel you have extra meaning by "allow compatibility". Would you elaborate on it a bit? Thanks!

rui-mo replied:
From my view, the velox4j does not intend to bind to any specific execution engine, am I correct?

zhztheplayer commented on java/README.md:
````diff
@@ -0,0 +1,201 @@
+
+```shell
+scripts/check.py header main --fix
+```
````
Same answer with https://github.com/facebookincubator/velox/pull/13186#discussion_r2135530532. Let me know if you have more comments on that.

rui-mo replied:
> Would you mind if we have an individual PR for enhancing the code documents after this is merged.

I would suggest we open a separate PR for docs.  Also, if some parts of the this PR are relatively independent, we may separate them as well.

zhztheplayer committed:
address comments

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,69 @@
+    return;
+  }
+  for (auto& child : rowBase->children()) {
+    if (child->size() > rowBase->size()) {
```
done

zhztheplayer committed:
fix include header

zhztheplayer committed:
remove unused file

zhztheplayer committed:
c++ code comments

zhztheplayer committed:
c++ code comments

zhztheplayer committed:
java comments

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/connector/ExternalStream.h:
```diff
@@ -0,0 +1,203 @@
+      void* context);
+
+ private:
+  const ObjectHandle esId_;
```
Would you document this variable?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/connector/ExternalStream.cpp:
```diff
@@ -0,0 +1,204 @@
+std::shared_ptr<ExternalStreamConnectorSplit>
+ExternalStreamConnectorSplit::create(const folly::dynamic& obj, void* context) {
+  const auto connectorId = obj["connectorId"].asString();
+  const auto esId = obj["esId"].asInt();
```
Perhaps just use `externalStreamId` to avoid unclear abbreviation.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/eval/Evaluation.h:
```diff
@@ -0,0 +1,56 @@
+
+  const facebook::velox::core::TypedExprPtr& expr() const;
+  const std::shared_ptr<const ConfigArray>& queryConfig() const;
+  const std::shared_ptr<const ConnectorConfigArray>& connectorConfig() const;
```
nit: add empty lines between methods.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/iterator/DownIterator.h:
```diff
@@ -0,0 +1,91 @@
+  /// the next row-vector from the stream.
+  facebook::velox::RowVectorPtr get();
+  /// Called to close the iterator.
+  void close();
```
nit: add empty line between methods. Please fix it through.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/iterator/UpIterator.cpp:
```diff
@@ -0,0 +1,16 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox4j/iterator/UpIterator.h"
```
The implementation is missing?

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,72 @@
+
+ArrowOptions makeOptions() {
+  ArrowOptions options;
+  options.timestampUnit = static_cast<TimestampUnit>(6);
```
I was aware of that. But haven't had a clear design on how to expose this to user as a configuration though. Let me open a ticket to follow up.

zhztheplayer committed:
address comment

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniCommon.cpp:
```diff
@@ -0,0 +1,111 @@
+
+const jint kJniVersion = JNI_VERSION_1_8;
+
+std::string jStringToCString(JNIEnv* env, jstring string) {
+  int32_t jlen, clen;
+  clen = env->GetStringUTFLength(string);
+  jlen = env->GetStringLength(string);
+  char buffer[clen];
+  env->GetStringUTFRegion(string, 0, jlen, buffer);
+  return {buffer, static_cast<uint64_t>(clen)};
+}
```
It's a nicer version. I am applying it.

jinchengchenghh replied:
https://github.com/apache/incubator-gluten/pull/9730

zhztheplayer committed:
pre-commit fix

zhztheplayer committed:
fixup

zhztheplayer committed:
pre-commit fix

zhztheplayer commented:
@rui-mo @assignUser @jinchengchenghh @lgbo-ustc @JkSelf 

Thank you for your review. I believe I have addressed the comments I could. Please feel free to continue review and I am seeking if we have opportunity to get the PR merged this week or so. :)

jinchengchenghh commented on .github/workflows/java-build-cpp.yml:
```diff
@@ -0,0 +1,56 @@
+on:
+  workflow_dispatch: {}
+  schedule:
+    - cron: 38 8 * * *
```
Could you receive the ci nitification if it failed

zhztheplayer replied:
These badges in `java/README.md` are for tracking the CI results (they may look broken now because the PR haven't merged):

https://github.com/zhztheplayer/velox/blob/1134924aff5e67ecc59844245c9f3d3756c021f0/java/README.md?plain=1#L3-L4

Another way is to find the results in the actions page: https://github.com/facebookincubator/velox/actions.

By the way, I made the C++ build test daily because I was concerned it could be too slow per PR. I can look into moving it back to PR wise once the CI time is proved to be acceptable with essential ccache optimizations.

jinchengchenghh commented on NOTICE.txt:
```diff
@@ -10,7 +10,10 @@ This product includes software from the QT project (BSD, 3-clause).
+This product includes software from the JniHelpers project (Apache 2.0).
+* https://github.com/spotify/JniHelpers/tree/master
+
+This product includes software from the The Arrow project.
```
the The -> the

zhztheplayer replied:
fixed

jinchengchenghh commented on .github/workflows/java-build-cpp.yml:
```diff
@@ -0,0 +1,56 @@
+            bash scripts/gha/ut-cpp/setup-centos7.sh
+            bash src/main/cpp/test.sh
+          "
+      - name: Save Ccache
```
Could we reuse the cache .github/workflows/linux-build-base.yml?

zhztheplayer replied:
Good question and it's hard to say yes at this moment because:

1. The build options may be different (Velox4J sets [these options](https://github.com/zhztheplayer/velox/blob/6bbd9de1333dd12333baa7f295e622e067851189/java/velox4j/src/main/cpp/CMakeLists.txt#L53-L61) constantly) which could result in cache misses
2. The source files may have different paths which could result in cache misses(I haven't verify this one but it likely be)

So I think we can keep the cache isolated for Java atm. And this practice will also help with the cache hit rate since the C++ build executions specifically for the Java bindings are anyway more similar to each other.

jinchengchenghh reviewed (commented):
Since this feature is a very big feature and verified in gluten with flink, I would suggest to review carefully for the code out of java folder, some small nits for java folder could be resolved in next iteration.

zhztheplayer committed:
fixup

jinchengchenghh reviewed (approved):
Thanks for you really good work!

assignUser reviewed (approved):
:+1: 

I have only one issue, that I don't think was discussed: the test parquet data adds about 30MB of binary data to the repo, which git handles badly. Can you not use the tpch connector to generate these dynamically? Or would that increase build times etc. too much?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniWrapper.cpp:
```diff
@@ -0,0 +1,471 @@
+
+jlong createEmptyBaseVector(JNIEnv* env, jobject javaThis, jstring typeJson) {
+  JNI_METHOD_START
+  // TODO Session memory pool.
```
Could you use a full sentence to describe what needs to be done in the future?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniWrapper.cpp:
```diff
@@ -0,0 +1,471 @@
+    jlong cSchema,
+    jlong cArray) {
+  JNI_METHOD_START
+  // TODO Session memory pool.
```
ditto

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniWrapper.cpp:
```diff
@@ -0,0 +1,471 @@
+  std::istringstream dataStream(decoded);
+  auto pool = session->memoryManager()->getVeloxPool(
+      "Decoding Memory Pool", memory::MemoryPool::Kind::kLeaf);
+  std::vector<ObjectHandle> vids{};
```
Do you mean vectorIds?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniWrapper.cpp:
```diff
@@ -0,0 +1,471 @@
+}
+} // namespace
+
+void JniWrapper::mapFields() {}
```
Do we have any TODO for this empty function?

zhztheplayer replied:
No need to add todos. They are just empty implementations. I added some comments.

rui-mo reviewed (commented):
Thanks. Added some nits, and could you please take a look at https://github.com/facebookincubator/velox/pull/13186#discussion_r2140831477 I added last time?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/arrow/Arrow.cpp:
```diff
@@ -0,0 +1,72 @@
+void slice(VectorPtr& in) {
+  auto* rowBase = in->as<RowVector>();
+  if (!rowBase) {
+    return;
```
I suppose the input of `fromBaseVectorToArrow` should all be row vector, do we need to handle a single-field vector?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/ResourceMap.h:
```diff
@@ -0,0 +1,100 @@
+
+  void erase(ResourceHandle moduleId) {
+    const std::lock_guard<std::mutex> lock(mtx_);
+    VELOX_CHECK(
```
nit: VELOX_CHECK_EQ

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/Session.h:
```diff
@@ -0,0 +1,46 @@
+
+#include <memory>
+
+#include "ObjectStore.h"
```
nit: use full include path.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/Session.h:
```diff
@@ -0,0 +1,46 @@
+#include <memory>
+
+#include "ObjectStore.h"
+#include "velox4j/memory/MemoryManager.h"
```
Do you need <>?

zhztheplayer replied:
"velox4j/..." is not a 3rd header. Or you meant something else?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/Session.h:
```diff
@@ -0,0 +1,46 @@
+
+namespace facebook::velox4j {
+
+/// A JNI session that is bound to a JniWrapper.
```
nit: use // for single-line comment.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cpp:
```diff
@@ -0,0 +1,423 @@
+  arrowAllocator_ = std::make_unique<ListenableMemoryAllocator>(
+      defaultMemoryAllocator().get(), listener_.get());
+  std::unordered_map<std::string, std::string> extraArbitratorConfigs;
+  ArbitratorFactoryRegister afr(listener_.get());
```
Perhaps just name it as `arbitratorFactoryRegister`.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cpp:
```diff
@@ -0,0 +1,423 @@
+      candidates_;
+};
+
+class ArbitratorFactoryRegister {
```
Find the `ListenableArbitrator` and `ArbitratorFactoryRegister` are key to the memory manager. Would you document them?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cpp:
```diff
@@ -0,0 +1,423 @@
+                    kMemoryReclaimMaxWaitMs,
+                    std::string(kDefaultMemoryReclaimMaxWaitMs))))
+                .count()) {}
+  std::string kind() const override {
```
nit: add empty line above.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.cpp:
```diff
@@ -0,0 +1,423 @@
+    return freeBytes;
+  }
+
+  const uint64_t memoryPoolInitialCapacity_; // FIXME: Unused.
```
Could you revise the TODO here?

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/Query.h:
```diff
@@ -0,0 +1,56 @@
+
+  const std::shared_ptr<const facebook::velox::core::PlanNode>& plan() const;
+  const std::shared_ptr<const ConfigArray>& queryConfig() const;
+  const std::shared_ptr<const ConnectorConfigArray>& connectorConfig() const;
```
nit: add empty line between.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/QueryExecutor.h:
```diff
@@ -0,0 +1,96 @@
+#include <memory>
+#include <vector>
+
+#include "Query.h"
```
nit: use full include path.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/QueryExecutor.h:
```diff
@@ -0,0 +1,96 @@
+  std::unique_ptr<SerialTaskStats> collectStats();
+
+ private:
+  State advance0(bool wait);
```
Perhaps rename as advanceInternal for readability.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/QueryExecutor.cpp:
```diff
@@ -0,0 +1,204 @@
+      cache::AsyncDataCache::getInstance(),
+      memoryManager_
+          ->getVeloxPool(
+              fmt::format("Query Memory Pool - EID {}", std::to_string(eid)),
```
Perhaps use full words for `EID` in message .

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/QueryExecutor.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  if (!task_->supportSerialExecutionMode()) {
+    VELOX_FAIL(
+        "Task doesn't support single threaded execution: " + task->toString());
```
Perhaps just VELOX_CHECK.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/query/QueryExecutor.cpp:
```diff
@@ -0,0 +1,204 @@
+SerialTask::~SerialTask() {
+  if (task_ != nullptr && task_->isRunning()) {
+    // FIXME: Calling .wait() may take no effect in single thread execution
+    //  mode.
```
Please start with TODO for the message.

zhztheplayer committed:
address nits

zhztheplayer committed:
pre-commit fix

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniCommon.cpp:
```diff
@@ -0,0 +1,113 @@
+    const std::string threadName =
+        fmt::format("Velox4J Native Thread {}", nextThreadId++);
+    std::vector<char> threadNameCStr(threadName.length() + 1);
+    std::strcpy(threadNameCStr.data(), threadName.data());
```
We are creating a C string (char*) here, how could string view be helpful? Am I missing something?

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/JniCommon.h:
```diff
@@ -0,0 +1,166 @@
+  SafeNativeArray& operator=(const SafeNativeArray&) = delete;
+  SafeNativeArray& operator=(SafeNativeArray&&) = delete;
+
+  const NativeArrayType elems() const {
```
Would you suggest `elements` or something?

rui-mo replied:
Sure, if it means `elements`, thanks.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/jni/StaticJniWrapper.cpp:
```diff
@@ -0,0 +1,428 @@
+}
+
+jstring
+deserializeAndSerializeVariant(JNIEnv* env, jobject javaThis, jstring json) {
```
This is an out-of-date code. Thanks for catching.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/ResourceMap.h:
```diff
@@ -0,0 +1,100 @@
+
+template <typename T, typename F>
+T safeCast(F f) {
+  VELOX_CHECK(sizeof(T) <= sizeof(F), "Vain safe casting");
```
I am removing the check. I was also aware of this but haven't come up with a good solution. This check is not necessary anyway.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/lifecycle/ResourceMap.h:
```diff
@@ -0,0 +1,100 @@
+
+template <typename T, typename F>
+T safeCast(F f) {
+  VELOX_CHECK(sizeof(T) <= sizeof(F), "Vain safe casting");
```
see https://github.com/facebookincubator/velox/pull/13186#discussion_r2156598963

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/memory/AllocationListener.cpp:
```diff
@@ -0,0 +1,64 @@
+  }
+  int64_t granted = reserve(diff);
+  if (granted == 0) {
+    return;
```
Good point. Fixed.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/memory/JavaAllocationListener.cpp:
```diff
@@ -0,0 +1,83 @@
+  } catch (const std::exception& ex) {
+    LOG(WARNING) << "Unable to destroy the global reference to the Java side "
+                    "allocation listener: "
+                 << ex.what();
```
This is in destructor where we don't usually throw exceptions.

zhztheplayer commented on java/velox4j/src/main/cpp/.clang-format:
Currently the C++ `format.sh` only mounts the CPP directory into docker:

https://github.com/zhztheplayer/velox/blob/74994c0fe6fadb482f91ceb01774a754b366ff3f/java/velox4j/scripts/gha/format/format.sh#L67C22-L68

So it won't see the parent clang-format file. I assume we can soon remove the script and migrate to pre-commit so maybe we just keep it for a while?

zhztheplayer committed:
fixup

zhztheplayer committed:
fixup

zhztheplayer committed:
remove gpg plugin

zhztheplayer committed:
download tpc-h data as a temporary solution to save the repo size

zhztheplayer committed:
remove binary parquet data

zhztheplayer committed:
nit

rui-mo reviewed (commented):
Thanks for iterating! Added some nits.

zhztheplayer commented:
@rui-mo 

Thanks. May you have other comments? I will address in one go tomorrow.

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/AllocationListener.h:
```diff
@@ -0,0 +1,86 @@
+  /// Be positive or negative. The returned value is the amount
+  /// of byte difference granted by the ceiling algorithm, which rounds
+  /// up the input bytes + current reserved bytes into the multiples
+  /// of block size.
```
nit: use //

rui-mo commented on java/velox4j/src/main/cpp/main/velox4j/memory/MemoryManager.h:
```diff
@@ -0,0 +1,68 @@
+  /// Destructs the members in this memory manager with a leakage-check
+  /// conducted.
+  /// If there are outstanding memory allocations (which means memory is leaked)
+  /// , returns false. Returns true otherwise.
```
nit: use //

rui-mo reviewed (commented):
Thanks! LGTM overall.

rui-mo commented:
> I temoprarily stored the data [here](https://github.com/zhztheplayer/tpc-data) for downloading.
> 
> There is now test utility [DownloadedTpchDataset](https://github.com/facebookincubator/velox/blob/1f8588fe72a76180056d75c02a0304bd4b97660a/java/velox4j/src/test/java/com/facebook/velox4j/test/dataset/tpch/DownloadedTpchDataset.java) responsible for downloading the dataset. It inherits from interface [TpchDataset](https://github.com/facebookincubator/velox/pull/13186/commits/1f8588fe72a76180056d75c02a0304bd4b97660a#diff-02874f4e043ceb0f5b51b7a0d71dc92c0a5df34ce542007716572d2e7f6468df) so we could switch to a local generator in the future by adding a new implementation of it.

Is there a way to avoid using a personal repository to download the test data? Is it possible to use Git LFS without making the Velox repository bloated?

https://git-lfs.com

zhztheplayer committed:
address change from https://github.com/facebookincubator/velox/pull/12993

zhztheplayer committed:
sync https://github.com/velox4j/velox4j/pull/242

zhztheplayer committed:
address comments

zhztheplayer commented on java/velox4j/src/main/java/com/facebook/velox4j/jni/LocalSession.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  @Override
+  public BaseVectors baseVectorOps() {
```
changed to `Operations`

zhztheplayer commented on java/velox4j/src/main/java/com/facebook/velox4j/jni/StaticJniApi.java:
```diff
@@ -0,0 +1,152 @@
+    return Serde.fromJson(json, ISerializable.class);
+  }
+
+  public Variant variantAsJava(VariantCo co) {
```
changed to `Cpp`

zhztheplayer commented on java/velox4j/src/main/java/com/facebook/velox4j/resource/Resources.java:
```diff
@@ -0,0 +1,184 @@
+    }
+    try {
+      zf.close();
+    } catch (final IOException e1) {
```
fixed

zhztheplayer commented on java/velox4j/src/main/java/com/facebook/velox4j/variant/MapValue.java:
```diff
@@ -0,0 +1,129 @@
+    return "MapValue{" + "map=" + map + '}';
+  }
+
+  public static class Entries {
```
Not really, it's exposed by API `getEntries`.

zhztheplayer commented on java/velox4j/src/main/cpp/test/velox4j/iterator/BlockingQueueTest.cpp:
```diff
@@ -0,0 +1,131 @@
+#include "velox4j/iterator/BlockingQueue.h"
+#include "velox4j/test/Init.h"
+
+namespace facebook::velox4j {
```
good point

zhztheplayer commented on java/velox4j/src/main/cpp/test/velox4j/test/Init.h:
```diff
@@ -0,0 +1,29 @@
+#include "velox4j/init/Init.h"
+
+namespace facebook::velox4j {
+inline void testingEnsureInitializedForSpark() {
```
Good idea while it needs a minor refactor against the CMake script. I'd follow-up on this.

zhztheplayer commented on java/velox4j/src/test/java/com/facebook/velox4j/test/ConfigTests.java:
```diff
@@ -0,0 +1,44 @@
+/*
```
I was intentionally avoiding naming anything with `util` / `Util` since they sound too general. This is more a personal practice but let's keep it until we found the rule is no longer sufficient for real development.

zhztheplayer commented on java/velox4j/src/main/cpp/main/velox4j/init/Init.cpp:
```diff
@@ -0,0 +1,138 @@
+
+namespace {
+/// A utility function to ensure the initialization code body, i.e., f(),
+/// is called once and only once when initialize(...) is being invoked.
```
As per [the documentation](https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md) the rule is 

```
For comments on headers for classes, functions, methods and member variables (the ones to be exposed as documentation), use /// (three slashes) at the front of each line. 
```

Am I getting something wrong here?

zhztheplayer committed:
style

zhztheplayer commented:
@rui-mo `git lfs` may be a good choice here, but my first impression is it will add a bit complexity for development as it requires an extra installation step of the git extension. This could cost user sometime to learn and fix the relevant issues.

Instead I am transferring the `zhztheplayer/tpc-data` repo to `velox4j/tpc-data`, I hope it could look less personalized in Velox's code base.

We should also eventually move to a local generator which is way more flexible than other solutions.

zhztheplayer committed:
move data to velox4j/tpc-data

zhztheplayer committed:
address https://github.com/facebookincubator/velox/pull/13885

zhztheplayer committed:
nit: slashes

zhztheplayer committed:
fixup

zhztheplayer committed:
https://github.com/facebookincubator/velox/pull/13920, https://github.com/facebookincubator/velox/pull/13959

zhztheplayer commented:
Hi @pedroerp @kgpai, this is ready for merge. Thanks!
